lib/auth.js

```diff
@@ -18,27 +18,31 @@ authenticator.getAccessToken = function(tenantId, spnAppPrincipalId, spnSymmetri
   var key = new Buffer(spnSymmetricKeyBase64, 'base64');
   var token = jwt.encode(payload, key);
   var data = {
-        grant_type: 'http://oauth.net/grant_type/jwt/1.0/bearer',
-        assertion: token,
-        resource: '00000002-0000-0000-c000-000000000000/directory.windows.net@' + tenantId
-         };
+    grant_type: 'http://oauth.net/grant_type/jwt/1.0/bearer',
+    assertion: token,
+    resource: '00000002-0000-0000-c000-000000000000/directory.windows.net@' + tenantId
+  };
 
   request.post('https://accounts.accesscontrol.windows.net/tokens/OAuth/2', { form: data }, function(e, resp, body) {
     if (e) return callback(e, null);
 
+    var parsedBody;
+
+    try {
+      parsedBody = JSON.parse(body);
+    } catch(e){
+      return callback(new Error('Failed to parse response from graph API'));

```

Comment:
It would be good to propagate (1) the contents of the original `e.message` and (2) the contents of `body` in the error. That way we could decide to log these [further up the stack](https://github.com/auth0/auth0-server/blob/master/lib/logs/serializers.js) (with a custom error name as well to be able to distinguish from other errors).

It might be very hard (impossible?) to troubleshoot any occurrences in prod without this info.

Reply:
It's worth doing the same with existing errors in other parts of the lib e.g. https://github.com/auth0/node-waad/blob/master/lib/waad.js#L42

Reply:
Also, if the request returns an error, this is already returned higher up `if (e) return callback(e, null)` so we do not have access to the original error message. If the body does not have PII then yes it will have to return this (otherwise we rely on the existing error handling in the `if (resp.statusCode != 200)` block below.

Reply:
> are we confident the contents of body will not expose any PII?

Unfortunately, no. What's more, we don't control it so it's impossible to guarantee that it doesn't contain PI/PII... However, this falls under "Information we receive from other sources" in the [Auth0 Privacy & Cookie Policy](https://auth0.com/privacy) so it is a legitimate use, with a couple of caveats:
1. the policy could be clearer about how we use this info for troubleshooting errors
1. we could do a lot better restricting access to error logs to those who need it

I believe these will get addressed as Auth0 matures as a cloud provider.

Having said that, attaching it to the error does not, by itself, expose the body anywhere where it shouldn't. :)